skfem/experimental/autodiff/helpers.py

```diff
@@ -0,0 +1,42 @@
+import jax.numpy as jnp
+
+from skfem import DiscreteField
+
+
+def dot(u, v):
+    return jnp.einsum('i...,i...', u, v)
+
+
+def ddot(u, v):
+    return jnp.einsum('ij...,ij...', u, v)
+
+
+def grad(u):
+    if isinstance(u, DiscreteField):
+        return u.grad
+    return u[1]
+
+
+def dd(u):
+    if isinstance(u, DiscreteField):
+        return u.hess
+    return u[4]
+
+
+def transpose(T):
+    return jnp.einsum('ij...->ji...', T)
+
+
+def mul(A, B):
+    if len(A.shape) == len(B.shape):
+        return jnp.einsum('ij...,jk...->ik...', A, B)
+    return jnp.einsum('ij...,j...->i...', A, B)
+
+
+def trace(T):
```

Comment:
Just a sidenote: the default NumPy function for the trace is fine here as the default axes are `0` and `1`, see https://numpy.org/doc/stable/reference/generated/numpy.trace.html.

```python
def trace(T):
    return jnp.trace(T, axis1=0, axis2=1)
```

Reply:
The only thing I wonder: In my tests like the one from my posted link above (https://github.com/adtzlr/matadi/discussions/11#discussioncomment-2554834) I had to sum the result of a functional over the trailing axes before taking the jacobian. Otherwise I got a full jacobian of every integration point of every element w.r.t. each other element. And I only want to have it quadrature-point-element wise. As you're using the `jvp` that *problem* seems to be solved?

Reply:
I need to look into this more carefully. I'm not exactly sure why the current approach works and I just managed to get the correct result by trying out jvp. @bhaveshshrimali  reports some slowness so there might be optimizations.